### ISSUE TITLE
Focus bug on text control inside the block settings sidebar (WP 5.4 RC2)

### ISSUE THREAD

**Issue by dianeco** [WordPress/gutenberg#20950]
Using WP 5.4 RC2, there's a problem of focus on text control (located inside the block settings sidebar) when selecting the field to edit it.

1. Insert a Heading block
2. Open the Block settings sidebar
3. Inside the HTML anchor field, type a word
4. Using your mouse, select all the word to edit it
5. The HTML anchor field loses the focus

The same issue happens with Additional CSS Class(es) field, Image dimensions field (Image block) ...
If you try to enter width and height values for an image, you end up editing the caption instead.

**Comment by jorgefilipecosta** [comment]
I was able to reproduce this issue on the last release of Gutenberg and WordPress 5.4 but not on WordPress 5.3.2, so it is a regression.

**Comment by jorgefilipecosta** [comment]
The issue is fixed on master and should be fixed on WordPress 5.4 RC4.

**Comment by jorgefilipecosta** [comment]
Thank you for reporting this issue @dianeco!

**Comment by dpkonofa** [comment]
I am still getting this on 5.4 release.

**Comment by dianeco** [comment]
@jorgefilipecosta It seems that the bug has reappeared in WP 5.4.1

**Comment by jorgefilipecosta** [comment]
It is not happening on Gutenberg master or the last version of Gutenberg but unfortunately, it seems it regressed on WP 5.4.1.

**Comment by JohnRDOrazio** [comment]
I'm having this problem now in WP 5.6.2. I have added a `TextControl` to the `InspectorControls` for my plugin's block, but when I try typing in the text input in the block's sidebar, as soon as I type one letter the input loses focus. I have to click the text input after every letter I type...
It is not happening instead in the `Advanced` section, for the `Additional CSS Classes` text input.

UPDATE: I finally found the cause of my issue. I was setting a key on the root element of my block with a dynamic function that incremented the value by 1 every time it was called, thinking I could generate a unique key this way for every element. However the increment function was incrementing every time state was set, so react was treating it as a new element and not the same element. Removing that key definition on the root element fixed everything.

**Comment by vidomarkas** [comment]
> I'm having this problem now in WP 5.6.2. I have added a `TextControl` to the `InspectorControls` for my plugin's block, but when I try typing in the text input in the block's sidebar, as soon as I type one letter the input loses focus. I have to click the text input after every letter I type... It is not happening instead in the `Advanced` section, for the `Additional CSS Classes` text input.

Same here. This issue still exists on wp 5.9.3

**Comment by KenjiMcAuliffe** [comment]
@vidomarkas I had the same issue when writing the editor code for a Gutenberg block. For some reason when my InspectorControls were broken out into their own function, rather than being written inside the "return" of the Edit function, this issue arose. Moving the InspectorControls back into the "return" resolved the issue. I have tried this for the NumberControl and the TextControl functional components.

**Comment by JohnRDOrazio** [comment]
I don't know what I'm doing wrong, but I'm still having this issue which now applies to all elements in my `InspectorControl` settings area. I have the InspectorControls in the return of the Edit function, so I don't believe that's the issue. I'm using `ServerSideRender`, I wonder if that has anything to do with it? Maybe the re-rendering of the block is influencing the focus on the controls? The Panels that the controls are contained in are also collapsing as soon as the block re-renders. And this makes it impossible to use any text controls that require typing...

The plugin code is in this repository: https://github.com/BibleGet-I-O/bibleget-wordpress/

The ServerSideRender callback function is here: https://github.com/BibleGet-I-O/bibleget-wordpress/blob/b052e81838fcd030c6c4183ed5390928deac32bc/bibleget-io.php#L728

The block code is here: https://github.com/BibleGet-I-O/bibleget-wordpress/blob/master/js/gutenberg.js

I also tried returning `setAttributes` in every `onchange` function callback instead of simply calling `setAttributes` but that didn't change anything.

I'm now up on WordPress 6.4 and Gutenberg 17.7.0

**Comment by djcowan** [comment]
@KenjiMcAuliffe ;  encountering similar issues with a functional component outside the Edit return.
the docs show the below example. However useState didn't appear to resolve the issue. useEffect ?? useRef? I don't know ima phpeeper <-- i like this. I should change my git handle.
```jsx
const MyTextControl = () => {
  const [ className, setClassName ] = useState( '' );

  return (
    <TextControl
      label="Additional CSS Class"
      value={ className }
      onChange={ ( value ) => setClassName( value ) }
    />
  );
};
```
I'm trying to reuse the same component in multiple locations, which is fine as long as I have multiple inline textControls within the return function. But guess the UI doesn't know where to return my change function to. This is probably unrelated to this thread. and should either be a new issue[?] or should be posted on stackOverflow where tech teens can flame me for not reading the docs.

https://github.com/user-attachments/assets/0018fdca-9e7f-4ed2-80bb-356310afedbe

**Comment by djcowan** [comment]
Okay. so I read the docs. or watched a video by some who had read the docs. And apparently, the issue relates to the lifecycle of react functions. each key press re/renders the component and therefor `const = myFunction` is a new `const = myFunction`. Reminds me of playing 'peek-a-boo' with a baby; something something... Object permanence. 
Moving myFunction 'outsite' of the component and feeding it with the props worked. Which is probably why this thread is closed and everyone involved is doing something productive.
